### PR TITLE
Feature - IJ-159 Wellbeing Services

### PR DESCRIPTION
**Pull request by IeuanSkinner**
### New Entities

I've introduced 2 new entities: wellbeing services and metrics services (which acts as a join table for wellbeing services and wellbeing metrics). The ERD for Wellbeing Assessments on LucidChart has been updated to reflect these table and can also be seen here:

Screenshot 2021-04-29 at 13 55 18

### Team Members

From a team members perspective there are now 2 new buttons on the dashboard, the first allows them to consider new wellbeing services which includes that services name, a descriptions, a link and connections between that service and the metrics.

Dashboard
Screenshot 2021-04-29 at 13 58 11

Wellbeing Services Index
Screenshot 2021-04-29 at 13 58 26

Wellbeing Services New/Edit Form
Screenshot 2021-04-29 at 13 58 46

The second button allows the team members to view the Wellbeing Metrics, currently I haven't added in the add/delete functionality as Include UK have told us these metrics are unlikely to change however they can inline edit them in case they feel the naming needs to be adjusted.

Wellbeing Metrics Index
Screenshot 2021-04-29 at 14 01 22

Wellbeing Metrics Inline Editing
Screenshot 2021-04-29 at 14 01 36

### Users

From the users perspective they can now click on the My Support button on the dashboard and they'll see an index page of all the wellbeing services which is using the exact same markup as the team members view (sans controls).
Screenshot 2021-04-29 at 14 02 30

When a user adjusts their score for a wellbeing metric this will now trigger the services that are linked to said metric to appear in the left-hand 'Your Services' section with a fade animation.
https://user-images.githubusercontent.com/6815945/116555314-261f3c80-a8f4-11eb-9a7f-7862bdcdff63.mov

### Additional Info

I've also added in a 'Coming Soon' controller that is linked to the buttons on the dashboard for which we aren't going to implement functionality as yet, when a user clicks these buttons now they are simply redirected back to the dashboard and shown a coming soon message.

Lastly **I'm aware there are UI improvements that can be made** but I've not been provided with any sort of reference for what this UI should look like nor are any of us experts on UI/UX (me especially!) so if those are the sorts of changes being suggested please do just go ahead and make them yourself.

### CODE REVIEW

File: app/controllers/users/coming_soon_controller.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Users
+  # app/controllers/users/coming_soon_controller.rb
+  class ComingSoonController < UsersApplicationController
+    def coming_soon
+      redirect_back(fallback_location: authenticated_user_root_path, notice: 'This functionality is coming soon.')
+    end
+  end
+end
```
Comment (phil-6) [inline app/controllers/users/coming_soon_controller.rb:8]: Does this warrant its own controller? Or can we just do it in the front-end?

File: app/views/team_members/wellbeing_services/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+          <%= f.text_field :name, class: 'form-control', required: true %>
+          <%= f.label :name %>
+        </div>
+        <div class="form-floating">
+          <%= f.text_area :description, class: 'form-control' %>
+          <%= f.label :description %>
+        </div>
+        <div class="form-floating">
+          <%= f.text_field :link, class: 'form-control', required: true %>
+          <%= f.label :link %>
```
Comment (phil-6) [inline app/views/team_members/wellbeing_services/_form.html.erb:17]: I think "Website" is more descriptive for the label.

File: app/views/team_members/wellbeing_services/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        </div>
+        <div class="container p-0">
+          <div class="row m-0">
+            <% wellbeing_metrics.each do |metric| %>
```
Comment (phil-6) [inline app/views/team_members/wellbeing_services/_form.html.erb:21]: Can we add a title or something here to explain what the buttons do. At first glance I don't think its not totally obvious.

File: app/views/team_members/wellbeing_services/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                                              class: 'btn btn-primary' %></div>
+              <div class="col-2"></div>
+            </div>
+
+            <% if @wellbeing_service.id.present? %>
+              <div class="row">
+                <div class="col-2"></div>
+                <div class="col-8">
+                  <%= link_to 'Remove', wellbeing_service_path(@wellbeing_service), method: :delete, class: 'btn btn-danger',
+                              data: { confirm: 'This will delete the wellbeing service. Are you sure?' } %>
```
Comment (phil-6) [inline app/views/team_members/wellbeing_services/_form.html.erb:45]: I prefer `btn-sm` for delete buttons / secondary actions on forms. Helps avoid accidental clicks.

File: app/assets/stylesheets/global.scss [hunk 5]
```diff
@@ -23,6 +23,10 @@ html, body {
   }
 }
 
+.v-hidden {
+  visibility: hidden !important;
+}
+
```
Comment (phil-6) [inline app/assets/stylesheets/global.scss:29]: could we not use bootstrap's built in `d-none` for this?

**Review by phil-6 — COMMENTED** [review]
Overall I'm really happy with this. Shining example of a PR, and brilliant, well considered feature. Few small comments but really impressive work as ever Ieuan!

File: app/controllers/users/coming_soon_controller.rb (hunk 1 above)
Comment (IeuanSkinner) [reply app/controllers/users/coming_soon_controller.rb:8]: 🤷‍♂️ this is the simplest solution, the buttons function like all the others except in this case they're just producing the 'Coming Soon' flash message. 

This means when it comes time to integrate that new functionality you just have to update the path for that button and you're good to go.

Comment (phil-6) [reply app/controllers/users/coming_soon_controller.rb:8]: I was thinking just add `disabled` to the `<a>` tag and add some text on the button to say "Coming Soon"

Comment (IeuanSkinner) [reply app/controllers/users/coming_soon_controller.rb:8]: I don't think `a` tags have a disabled attribute.

File: app/views/team_members/wellbeing_services/_form.html.erb (hunk 3 above)
Comment (IeuanSkinner) [reply app/views/team_members/wellbeing_services/_form.html.erb:21]: If you've something in your mind feel free to implement.

File: app/controllers/users/coming_soon_controller.rb (hunk 1 above)
Comment (phil-6) [reply app/controllers/users/coming_soon_controller.rb:8]: Sorry, buttons: https://getbootstrap.com/docs/4.0/components/buttons/#disabled-state

File: app/assets/stylesheets/global.scss (hunk 5 above)
Comment (IeuanSkinner) [reply app/assets/stylesheets/global.scss:29]: Nope, we want the dummy button to occupy the same dimensions as though it were a real button and you lose that behaviour with `display: none`

**Review by AlexW555 — APPROVED** [review]
(Oh apologises, I already looked over this but I guess I closed the tab before I submitted)

Looks really good! I love the toggleable switches. Tested through the functionality and works perfectly!

**Comment by garethdandrews** [comment]
On smaller displays the metrics get cut off, might it be worth making the container a bit bigger for this one form?
Or put each metric on a new line.

Screenshot from 2021-05-11 14-55-03

Screenshot from 2021-05-11 14-55-46

**Comment by garethdandrews** [comment]
This is all looking really good! Really enjoying the in-line editing and fade animations.
Should team-members be able to add a new wellbeing metric? If yes I don't mind making a ticket and bashing it out quickly.

**Comment by garethdandrews** [comment]
> > This is all looking really good! Really enjoying the in-line editing and fade animations.
> > Should team-members be able to add a new wellbeing metric? If yes I don't mind making a ticket and bashing it out quickly.
> 
> @garethdandrews; The only issue that exists with this at present is how do you handle metrics being added and removed over time, I don't think there's anything we've written thus far that wouldn't be able to cope with so we probably could add that functionality but some experimenting would need to be done to be sure and it's not something Include UK have asked for.
> 
> @phil-6 thoughts?

Can probably just leave it for now then :+1:

**Comment by IeuanSkinner** [comment]
@garethdandrews; 

Updates to the wellbeing service form illustrated below.

Desktop:
Screenshot 2021-05-11 at 16 58 33

Mobile:
Screenshot 2021-05-11 at 16 58 13

**Pull request merged**